### PR TITLE
Add reference site export & import commands

### LINKED ISSUE #4571: Build reference test site
We should work on a reference test site that includes typical website content and try to use or touch most or all of the functionality of the AMP plugin as a real-world website would.

This will either simplify or be a requirement for the following features:

- Add performance metrics to CI [[#4569](https://github.com/ampproject/amp-wp/issues/4569)]
- Add unified sanitizer smoke test [[4549](https://github.com/ampproject/amp-wp/issues/4549)]
- Add automated visual regression testing [[4671](https://github.com/ampproject/amp-wp/issues/4671)]

### PR DESCRIPTION
**Pull request by schlessera**
## Summary

This PR adds two WP-CLI commands:
- `wp amp reference-site import <file> [--empty-content] [--empty-uploads] [--empty-extensions] [--empty-options] [--skip-site-meta]`
- `wp amp reference-site export <file> [--force]`

Reference site definitions can be created via the export command and reused for various purposes via the import command.

The main use case is to use the reference sites as a clearly defined starting point for performance & fidelity tests.

Media files are uploaded to and downloaded from Google Storage in the `ampwp_reference_sites_media_files` bucket. For this to work, the Google Cloud CLI tools have to be installed and the user needs to be logged in and working on the correct GCP project.

Note: This is a preliminary version for early committing into `develop` branch with the plan to refine and add tests in separate PRs down the road.

Part of #4571

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: tests/php/src/Cli/Export/ExportActivePlugins.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	 * List of plugin slugs to exclude.
+	 *
+	 * @var string[]
+	 */
+	const EXCLUDED_PLUGINS = [
+		'airplane-mode',
+		'amp',
+		'query-monitor',
+		'redis-cache',
+		'wp-rocket',
```
Comment (westonruter) [inline tests/php/src/Cli/Export/ExportActivePlugins.php]: Why these plugins specifically?

File: tests/php/src/Cli/Export/ExportActivePlugins.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	/**
+	 * Get the list of currently active plugins.
+	 *
+	 * @return string[] Array of currently active plugins.
+	 */
+	private function get_active_plugins() {
+		return array_map( static function ( $plugin ) {
+			$filename = basename( dirname( $plugin ) );
```
Comment (westonruter) [inline tests/php/src/Cli/Export/ExportActivePlugins.php]: What about a `$plugin` which is just a file and not containing a directory? Would it be better to use:

```suggestion
			$filename = strtok( $plugin, '/' );
```

File: tests/php/src/Cli/Export/ExportOptions.php [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+	 * @return ExportResult Adapted export result.
+	 */
+	public function process( ExportResult $export_result ) {
+		$option_keys = array_filter(
+			$this->get_option_keys(),
+			[ $this, 'skip_excluded_options' ]
+		);
+
+		$options = $this->fetch_options( $option_keys );
+		$options = array_merge( $options, $this->fetch_theme_mods() );
```
Comment (westonruter) [inline tests/php/src/Cli/Export/ExportOptions.php]: Why are the options being merged with the theme mods? How will these be properly re-imported as theme mods? Themes can have their own custom theme mods that aren't among the fixed set that core supports (e.g. colors). To properly handle theme mods, I think they should be exported separately and then have a separate importer that will copy them as-is to the theme mods, with special handling for the select few that need additional processing (e.g. custom logo).

File: tests/php/src/Cli/Export/ExportOptions.php [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+	/**
+	 * Get the attachment URL from an attachent ID.
+	 *
+	 * @param string|int $attachment_id ID of the attachment post to get the URL
+	 *                                  for.
+	 * @return string|false Attachment URL, or false if none.
+	 */
+	private function get_attachment_url_from_attachment_id( $attachment_id ) {
+		return wp_get_attachment_url( $attachment_id );
+	}
```
Comment (westonruter) [inline tests/php/src/Cli/Export/ExportOptions.php]: Why is this method needed? Seems like a verbose alias.

File: tests/php/src/Cli/Export/ExportOptions.php [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+			return true;
+		}
+
+		return self::OPTION_DEFAULTS[ $option_key ] !== $option_value;
+	}
+
+	/**
+	 * Get the post title from a post ID.
+	 *
+	 * @param string|int $post_id ID of the post to get the title from.
```
Comment (westonruter) [inline tests/php/src/Cli/Export/ExportOptions.php]:
```suggestion
	 * @param int $post_id ID of the post to get the title from.
```

File: tests/php/src/Cli/Export/ExportWidgets.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+		$exported_widgets = [];
+
+		foreach ( wp_get_sidebars_widgets() as $sidebar => $widgets ) {
+			$exported_widgets[ $sidebar ] = [];
+			foreach ( $widgets as $widget_instance_id ) {
+				// Get id_base (remove -# from end) and instance ID number.
+				$id_base = preg_replace( '/-[0-9]+$/', '', $widget_instance_id );
+				$instance = preg_replace( '/^' . preg_quote( $id_base, '/' ) . '-/', '', $widget_instance_id );
+				$widget_options = get_option( 'widget_' . $id_base, [] );
+				$exported_widgets[ $sidebar ][ $widget_instance_id ] = $widget_options[ (int) $instance ];
```
Comment (westonruter) [inline tests/php/src/Cli/Export/ExportWidgets.php]: While this will work for most sites now, the option name used to store a `WP_Widget`'s settings is not guaranteed to include the `$id_base`. The actual option name is stored in `WP_Widget::$option_name`. There is also a `\WP_Widget::$alt_option_name`! Lastly, widgets may not use `WP_Widget` at all. For example: https://gist.github.com/westonruter/7141599

**Review by westonruter — CHANGES_REQUESTED** [review]

File: tests/php/src/Cli/Export/ExportActivePlugins.php (hunk 1 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportActivePlugins.php]: It's a set of development plugins I'd happen to have installed and want to exclude. This list will probably further grow, but this was what solved it on my end to have a clean list of actual required plugins.

File: tests/php/src/Cli/Export/ExportOptions.php (hunk 3 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportOptions.php]: That was what the starter sites code from Astra was using which I used as initial test environment. But I agree, it makes more sense to split this out.

File: tests/php/src/Cli/Export/ExportOptions.php (hunk 4 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportOptions.php]: For most of these "adaptation" methods, they started out as basically aliases but then grew further and further as I fixed bugs and edge cases. This one happens to be a WP function that is properly built, I guess?

Happy to remove this indirection, though.

File: tests/php/src/Cli/Export/ExportWidgets.php (hunk 6 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportWidgets.php]: Yes, this was mainly a shortcut based what I assumed would be what we'd encounter while exporting the sites we'd just build.

I'll revisit this exporter to provide more complete support.

File: tests/php/src/Cli/Export/ExportActivePlugins.php (hunk 2 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportActivePlugins.php]: Yes, will adapt.

File: tests/php/src/Cli/Export/ExportWidgets.php (hunk 6 above)
Comment (westonruter) [reply tests/php/src/Cli/Export/ExportWidgets.php]: Probably only need to address it when it becomes an issue.

Comment (schlessera) [reply tests/php/src/Cli/Export/ExportWidgets.php]: That was the initial plan. 😀
Let's stick to that then!

File: tests/php/src/Cli/Export/ExportOptions.php (hunk 5 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportOptions.php]: We're talking about WordPress here... But I can cast to an integer when I call this method to enforce an int.

File: tests/php/src/Cli/Export/ExportOptions.php (hunk 3 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportOptions.php]: Resolved via https://github.com/ampproject/amp-wp/pull/5162/commits/2cd9e326dc2649b8317e27c65f0827aaa992a420

File: tests/php/src/Cli/Export/ExportActivePlugins.php (hunk 2 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportActivePlugins.php]: Resolved via https://github.com/ampproject/amp-wp/pull/5162/commits/806df0c4c5832707bf0d80a59d8494e5ea32feb4

File: tests/php/src/Cli/Export/ExportOptions.php (hunk 5 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportOptions.php]: Resolved via https://github.com/ampproject/amp-wp/pull/5162/commits/05d9be2779751fc870d8addacae5b3d09347d64b

File: tests/php/src/Cli/Export/ExportOptions.php (hunk 4 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportOptions.php]: Resolved via https://github.com/ampproject/amp-wp/pull/5162/commits/df2c3f8944bc577a623b299525c150022ba19f46

File: tests/php/src/Cli/Export/ExportActivePlugins.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+use AmpProject\AmpWP\Tests\Cli\ExportStep;
+
+final class ExportActivePlugins implements ExportStep {
+
+	/**
+	 * List of plugin slugs to exclude.
+	 *
+	 * @var string[]
+	 */
+	const EXCLUDED_PLUGINS = [
```
Comment (pierlon) [inline tests/php/src/Cli/Export/ExportActivePlugins.php:19]: I think this would be better suited in a config/env file.

File: tests/php/src/Cli/Export/ExportResult.php [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+	 * @return string Target path of the site definition file.
+	 */
+	public function get_target_path() {
+		return $this->target_path;
+	}
+
+	/**
+	 * Get the site name to use.
+	 *
+	 * @var string Site name to use.
```
Comment (pierlon) [inline tests/php/src/Cli/Export/ExportResult.php]:
```suggestion
	 * @return string Site name to use.
```

File: tests/php/src/Cli/Export/ExportThemeMods.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		unset( $theme_mods['custom_css_post_id'] );
+
+		return $theme_mods;
+	}
+
+	/**
+	 * Translate from menu_id into menu_slug.
+	 *
+	 * @param array $nav_menu_locations Associative array of nav menu locations.
+	 */
```
Comment (pierlon) [inline tests/php/src/Cli/Export/ExportThemeMods.php:75]:
```suggestion
	 * @return array Associative array of translated menu locations.
	 */
```

File: tests/php/src/Cli/Import/ActivatePlugin.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+	 *             Returns -1 for failure.
+	 */
+	public function process() {
+		WP_CLI::log(
+			WP_CLI::colorize(
+				"Installing and activating plugin %G'{$this->plugin}'%n..."
+			)
+		);
+
+		WP_CLI::runcommand( "plugin install {$this->plugin} --activate" );
```
Comment (pierlon) [inline tests/php/src/Cli/Import/ActivatePlugin.php]: What would happen if a plugin fails to install? Would it still return `1`?

File: tests/php/src/Cli/Import/ImportCustomizerSettings.php [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+		$this->settings = $settings;
+	}
+
+	/**
+	 * Process the step.
+	 *
+	 * @return int Number of items that were successfully processed.
+	 *             Returns -1 for failure.
+	 */
+	public function process() {
```
Comment (pierlon) [inline tests/php/src/Cli/Import/ImportCustomizerSettings.php:38]: Method isn't returning anything.

File: tests/php/src/Cli/Import/ImportOptions.php [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+		$this->options = $options;
+	}
+
+	/**
+	 * Process the step.
+	 *
+	 * @return int Number of items that were successfully processed.
+	 *             Returns -1 for failure.
+	 */
+	public function process() {
```
Comment (pierlon) [inline tests/php/src/Cli/Import/ImportOptions.php:38]: Method isn't returning anything.

File: tests/php/src/Cli/Import/ImportThemeMods.php [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+		$this->theme_mods = $theme_mods;
+	}
+
+	/**
+	 * Process the step.
+	 *
+	 * @return int Number of items that were successfully processed.
+	 *             Returns -1 for failure.
+	 */
+	public function process() {
```
Comment (pierlon) [inline tests/php/src/Cli/Import/ImportThemeMods.php:38]: Method isn't returning anything.

File: tests/php/src/Cli/ReferenceSiteImportCommand.php [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+			$this->empty_site( $empty_uploads );
+		}
+
+		if ( $empty_options ) {
+			$this->empty_options();
+		}
+
+		WP_CLI::log(
+			'Importing reference site: '
+			. WP_CLI::colorize( "%y{$site_definition->get_name()}%n (%gv{$site_definition->get_version()}%n)" )
```
Comment (pierlon) [inline tests/php/src/Cli/ReferenceSiteImportCommand.php:124]: `$site_definition` will be undefined if the try/catch above it's declared in throws an error.

File: tests/php/src/Cli/ReferenceSiteImporter.php [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+	}
+
+	/**
+	 * Check whether the provided URL is pointing to Google Storage.
+	 *
+	 * @param string $url URL to check.
+	 * @return bool Whether this is a Google Storage URL.
+	 */
+	private function is_google_storage_url( $url ) {
+		return 1 === preg_match( '#^gs://#', $url );
```
Comment (pierlon) [inline tests/php/src/Cli/ReferenceSiteImporter.php]: `preg_match` seems like a bit of overkill for this. Something like `strncmp` should do.like in: https://github.com/ampproject/amp-wp/blob/1f1ebeef03862e7a72716f182dc761e9690e732e/tests/php/src/Cli/Export/ExportWxrFile.php#L81

File: .gitignore [hunk 16]
```diff
@@ -18,3 +18,4 @@ built
 /lib/*/composer.lock
 /phpcs.xml
 /phpunit.xml
+*.sql
```
Comment (pierlon) [inline .gitignore]: Is this still relevant? If so maybe the path should be scoped to where the SQL files would be placed.

**Review by pierlon — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 16 above)
Comment (schlessera) [reply .gitignore]: I've found that I'm working with DB imports/exports when creating/editing a reference site, so adding this seems to make sense.

I'll scope it to the root folder.

File: tests/php/src/Cli/ReferenceSiteImportCommand.php (hunk 14 above)
Comment (schlessera) [reply tests/php/src/Cli/ReferenceSiteImportCommand.php:124]: The `WP_CLI::error()` ends with an `exit()` and stops the process.

File: tests/php/src/Cli/Import/ActivatePlugin.php (hunk 10 above)
Comment (schlessera) [reply tests/php/src/Cli/Import/ActivatePlugin.php]: Resolved via https://github.com/ampproject/amp-wp/pull/5162/commits/b9cc1ec30a41059cb671c0347aac1ce73059059a

File: tests/php/src/Cli/Import/ImportCustomizerSettings.php (hunk 11 above)
Comment (schlessera) [reply tests/php/src/Cli/Import/ImportCustomizerSettings.php:38]: Resolved via https://github.com/ampproject/amp-wp/pull/5162/commits/b9cc1ec30a41059cb671c0347aac1ce73059059a

File: tests/php/src/Cli/Import/ImportOptions.php (hunk 12 above)
Comment (schlessera) [reply tests/php/src/Cli/Import/ImportOptions.php:38]: Resolved via https://github.com/ampproject/amp-wp/pull/5162/commits/b9cc1ec30a41059cb671c0347aac1ce73059059a

File: tests/php/src/Cli/Import/ImportThemeMods.php (hunk 13 above)
Comment (schlessera) [reply tests/php/src/Cli/Import/ImportThemeMods.php:38]: Resolved via https://github.com/ampproject/amp-wp/pull/5162/commits/b9cc1ec30a41059cb671c0347aac1ce73059059a

File: tests/php/src/Cli/ReferenceSiteImporter.php (hunk 15 above)
Comment (schlessera) [reply tests/php/src/Cli/ReferenceSiteImporter.php]: https://github.com/ampproject/amp-wp/pull/5162/commits/31d773146f249e2a779b8c22167f4a9acc4752e6

File: .gitignore (hunk 16 above)
Comment (schlessera) [reply .gitignore]: Resolved via https://github.com/ampproject/amp-wp/pull/5162/commits/686e6c677837563924459a9c786fd96586c1768f

File: tests/php/src/Cli/Export/ExportActivePlugins.php (hunk 7 above)
Comment (schlessera) [reply tests/php/src/Cli/Export/ExportActivePlugins.php:19]: Resolved via https://github.com/ampproject/amp-wp/pull/5162/commits/5dec96eee8bd6972f9284c6eac2e95bd778b642b :

Added an environment variable `AMP_REF_SITE_EXCLUDED_PLUGINS`

**Comment by schlessera** [comment]
@westonruter Provided the tests pass and there's no major blocker for you, could we have this merged so I can reuse this in multiple other places?

More details about current state:
- Importing should work, and even work without being authenticated with Google Storage. The command I use is this:
	```
	wp amp reference-site import --empty-uploads --empty-extensions --empty-options --empty-content travel-blog
	```
- Exporting is generally working, however there's a few details around Google Storage URLs and authentication that need fine-tuning. I don't want this to block the use of the importer further, though, and reference site can also be built manually pretty easily for simpler use cases like the E2E scenario we discussed.

File: tests/php/src/Cli/Export/ExportOptions.php [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+	private function fetch_options( $keys ) {
+		global $wpdb;
+
+		if ( empty( $keys ) ) {
+			return [];
+		}
+
+		$query = sprintf(
+			'SELECT `option_name`, `option_value` FROM `%s` WHERE `option_name` IN (%s);',
+			esc_sql( $wpdb->options ),
```
Comment (westonruter) [inline tests/php/src/Cli/Export/ExportOptions.php:205]: This is escaped here it isn't above in the `get_option_keys` call.

Comment (westonruter) [reply tests/php/src/Cli/Export/ExportOptions.php:205]: It doesn't seem necessary since normally the `$wpdb->options` is passed into SQL as-is.

File: tests/php/src/Cli/Export/ExportOptions.php [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+			implode(
+				',',
+				array_map(
+					static function ( $key ) {
+						return "'" . esc_sql( $key ) . "'"; },
```
Comment (westonruter) [inline tests/php/src/Cli/Export/ExportOptions.php:210]: Minor point, but why not:

```suggestion
						return $wpdb->prepare( '%s', $key ),
```

File: tests/php/src/Cli/Export/ExportOptions.php [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+	 * @return string Post title.
+	 */
+	private function get_page_title_from_post_id( $post_id ) {
+		$post = get_post( $post_id );
+
+		if ( ! $post instanceof WP_Post || empty( $post->post_title ) ) {
+			return $post_id;
+		}
+
+		return $post->post_title;
```
Comment (westonruter) [inline tests/php/src/Cli/Export/ExportOptions.php:302]: Maybe `post_name` would make more sense?

File: tests/php/src/Cli/Import/ImportOptions.php [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+	 * Update option pointing to a page.
+	 *
+	 * This adapts the page ID as needed.
+	 *
+	 * @param string $key   Option key.
+	 * @param mixed  $value Option value.
+	 * @return bool Whether the update was successful.
+	 */
+	private function update_page_id_option( $key, $value ) {
+		$page = get_page_by_title( $value );
```
Comment (westonruter) [inline tests/php/src/Cli/Import/ImportOptions.php:101]: Per above above using `post_name`, here perhaps it would make sense to use `get_page_by_path()`

File: tests/php/src/Cli/Import/ImportWxrFile.php [hunk 21]
```diff
@@ -0,0 +1,348 @@
...
+use WP_User;
+
+final class ImportWxrFile implements ImportStep {
+
+	/**
+	 * Regular expression pattern to use for detecting upload URLs in post content.
+	 *
+	 * @var string
+	 */
+	const UPLOAD_URL_PATTERN = '#/wp-content/uploads/(\d+/\d+/.*?\.\w+)#i';
```
Comment (westonruter) [inline tests/php/src/Cli/Import/ImportWxrFile.php:25]: Note that this folder structure won't always be present, such as when this option is turned off:

image

Also, `wp-content` may not be where the content is.

File: tests/php/src/Cli/ReferenceSiteImportCommand.php [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+		);
+
+		$plugins = array_filter(
+			$plugins,
+			static function ( $plugin ) {
+				return ! in_array( $plugin, [ 'amp', 'wordpress-importer' ], true );
+			}
+		);
+
+		WP_CLI::runcommand( 'plugin delete ' . implode( ' ', $plugins ) );
```
Comment (westonruter) [inline tests/php/src/Cli/ReferenceSiteImportCommand.php:278]: Perhaps safer to deactivate rather than delete?

```suggestion
		WP_CLI::runcommand( 'plugin deactivate ' . implode( ' ', $plugins ) );
```

File: tests/php/src/Cli/ReferenceSiteImporter.php [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+	 *
+	 * @param string $string The string to check.
+	 * @return bool Whether the string is an image url or not.
+	 */
+	public static function is_image_url( $string = '' ) {
+		if ( ! is_string( $string ) ) {
+			return false;
+		}
+
+		if ( ! preg_match( '/\.(jpg|jpeg|svg|png|gif)$/i', $string ) ) {
```
Comment (westonruter) [inline tests/php/src/Cli/ReferenceSiteImporter.php:132]: Could also include `webp` and now `avif`

**Review by westonruter — APPROVED** [review]
